fix: Refactor filters parsing

coderbirju opened the pull request:
Issue #, if available:

*Description of changes:*
This commit adds filter types and parsing capabilities to easily fetch these for our containers list and events APIs.

Parsing logic has been ported over from https://github.com/moby/moby/blob/5f0d6731eb015c8e46b2ae9bb1803d005f2513be/api/types/filters/parse.go


*Testing done:*



- [ ] I've reviewed the guidance in CONTRIBUTING.md


#### License Acceptance

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

Shubhranshu153 commented on api/types/filter_types.go:
```diff
@@ -0,0 +1,145 @@
+}
+
+// KeyValuePair are used to initialize a new Args.
+type KeyValuePair struct {
```
can we name it more connected to the filters like filter type and filter value.

Shubhranshu153 commented on api/types/filter_types.go:
```diff
@@ -0,0 +1,145 @@
+
+// Args stores a mapping of keys to a set of multiple values.
+type Args struct {
+	fields map[string]map[string]bool
```
fields -> filter list

Shubhranshu153 commented on api/types/filter_types.go:
```diff
@@ -0,0 +1,145 @@
+}
+
+// Keys returns all the keys in list of Args.
+func (args Args) Keys() []string {
```
also can it be private?

Shubhranshu153 commented on api/types/filter_types.go:
```diff
@@ -0,0 +1,145 @@
+}
+
+// Get returns the list of values associated with the key.
+func (args Args) Get(key string) []string {
```
private function?

subashkotha commented on api/types/filter_types.go:
```diff
@@ -0,0 +1,145 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package types
```
Is type appropriate package for filters? This contains utility functions unlike other files under type which just have definitions.

coderbirju replied:
The Filters type are a unique case because we need these methods to convert them to our desired legacy format whenever we use them, I think this was the most appropriate place I could put the methods

coderbirju commented:
- Updated with better names, removed export of non necessary functions.
- Left one of the older e2e tests as it is to ensure that old format is parsed properly, new format is tested in another testcase
- ParseFilterArgs now returns Filters object and in conjuction with ToLegacyFormat() method we can get the desired format

coderbirju commented on api/types/filter_types.go:
```diff
@@ -0,0 +1,145 @@
+	return result, nil
+}
+
+func NerdctlFiltersFromAPIFilters(query url.Values) ([]string, error) {
```
moved this back to the list.go

Shubhranshu153 commented:
what if nerdctl filter doesnt support a particular docker filter.

swagatbora90 commented on e2e/tests/container_list.go:
```diff
@@ -226,7 +226,7 @@ func ContainerList(opt *option.Option) {
 			Expect(err).Should(BeNil())
 			defer res.Body.Close()
-			errorMsg := fmt.Sprintf("invalid query parameter \\\"filters\\\": %s", fmt.Errorf("invalid character 'i' looking for beginning of value"))
+			errorMsg := fmt.Sprintf("invalid query parameter \\\"filters\\\": %s", fmt.Errorf("error parsing filters: invalid character 'i' looking for beginning of value"))
```
Can you also add a e2e tests with the new filter format?

coderbirju replied:
added, https://github.com/runfinch/finch-daemon/blob/a3b76dd5428f552f73b69fd5ae93a3738eb77120/e2e/tests/container_list.go#L232
